crates/sp-domains-fraud-proof/src/host_functions.rs

```diff
@@ -355,6 +355,13 @@ where
                 .map(|is_inherent| {
                     FraudProofVerificationInfoResponse::InherentExtrinsicCheck(is_inherent)
                 }),
+            FraudProofVerificationInfoRequest::ExtrinsicDecodableCheck(opaque_extrinsic) => {
+                let encoded_extrinsic = opaque_extrinsic.encode();
+                Some(FraudProofVerificationInfoResponse::ExtrinsicDecodableCheck(
+                    <DomainBlock as BlockT>::Extrinsic::decode(&mut encoded_extrinsic.as_slice())
+                        .is_ok(),
+                ))
```

Comment:
Ideally, we should replace `T::DomainHeader` with `T::DomainBlock` in the pallet-domains and use `<T::DomainBlock as BlockT>::Extrinsic::decode` to perform the check to avoid involving host function as possible, but it will require a significant refactoring since `T::DomainHeader` is used so many places. Plz add some comments here in case we can do something in the future.

Reply:
> Ideally, we should replace T::DomainHeader with T::DomainBlock in the pallet-domains and use <T::DomainBlock as BlockT>::Extrinsic::decode

Nope, that is incorrect. Domain's UncheckedExtrinsic would vary depending on the type of Extrinsic domain would use. For now we can already use Consensus block extrinsinc to decode such but this is also not future proof since the extrinsic type may very. 

introducing T::DomainBlock is also a false assumption since each domain will have their own set of concrete definition of Domain block.

The future way of decoding such domain extrinsic is use the domain specific runtime to try and decode such extrinsic and it is totally valid to use Hostfunction for this purpose so that we are not limiting ourselves with the implementation unnecessarily and us introducing the host function is purely to achive this indetermined domain specific type

Reply:
>For now we can already use Consensus block extrinsinc to decode such but this is also not future proof since the extrinsic type may very.

This sounds like an issue already, could you link me to the code?

>introducing T::DomainBlock is also a false assumption since each domain will have their own set of concrete definition of Domain block.

With a closer look, I see the issue now, on the client side what we are use for `DomainBlock` is actually `domain_runtime_primitives::opaque::Block` with `OpaqueExtrinsic`. So the decodable check doesn't work at all since it just decodes the bundle's `OpaqueExtrinsic` into `OpaqueExtrinsic`:
https://github.com/subspace/subspace/blob/8f53e8bba65dd17c7f7531b47f1f981d086b5bde/domains/client/block-preprocessor/src/lib.rs#L266-L267

I think we need to introduce a domain runtime API to take `OpaqueExtrinsic` and decode it to `<Block as BlockT>::Extrinsic` (in the runtime API the `Block` is the concrete runtime Block), and use it in both the above check and the fraud proof verification cc @rahulksnv

Reply:
As someone new to the domains code, I found the whole structure confusing/error prone: like in this case, usage of `CBlock/Block/DomainBlock`

I would implement this using polymorphism/dynamic dispatch rather than generics. I feel that would give a better separation of L1/consensus <-> L2/execution/domains

Reply:
> This sounds like an issue already, could you link me to the code?

We do not have any code that does this but potential sloution which I do not like and should completely disregard this approach

> I think we need to introduce a domain runtime API to take OpaqueExtrinsic and decode it to <Block as BlockT>::Extrinsic

Yeap. That is exactly what I'm proposing here :)

Reply:
>usage of CBlock/Block/DomainBlock

This is inherited from substrate, if we just keep `CBlock` and `DomainBlock` it may be more clear. Nevertheless, even the `block-preprocessor`'s decodable check is problematic we need to fix it with:
>I think we need to introduce a domain runtime API to take OpaqueExtrinsic and decode it to <Block as BlockT>::Extrinsic (in the runtime API the Block is the concrete runtime Block), and use it in both the above check and the fraud proof verification

Reply:
@vedhavyas When the stateless runtime is instantiated, the `DomainBlock` is already fixed (e.g) https://github.com/subspace/subspace/blob/main/crates/sp-domains-fraud-proof/src/host_functions.rs#L228. Beyond that, the stateless runtime takes the extrinsic already in the domain specific extrinsic format `<DomainBlock as BlockT>::Extrinsic`

May be I am missing something?

Reply:
@rahulksnv `DomainBlock` there is an opaque one defined [here](https://github.com/subspace/subspace/blob/main/domains/primitives/runtime/src/lib.rs#L130) we use as a place holder for client due to how substrate defines them.

What we actually want to decode to is not opaque one but rather the actual extrinsic definition defined [here](https://github.com/subspace/subspace/blob/main/domains/runtime/evm/src/lib.rs#L102) which only the runtime would know. So we want to decode this extrinsic to that second type

Reply:
Thanks @vedhavyas, PTAL. Will add tests separately if this looks inline with what you had in mind